pkg/controller/v1alpha1/localmodelnode/controller.go

```diff
@@ -192,60 +162,127 @@ func (c *LocalModelNodeReconciler) getContainerSpecForStorageUri(ctx context.Con
 // Update the status of the LocalModelNode CR
 func (c *LocalModelNodeReconciler) downloadModels(ctx context.Context, localModelNode *v1alpha1api.LocalModelNode) error {
 	newStatus := map[string]v1alpha1api.ModelStatus{}
 	for _, modelInfo := range localModelNode.Spec.LocalModels {
-		if status, ok := localModelNode.Status.ModelStatus[modelInfo.ModelName]; ok {
-			if status == v1alpha1api.ModelDownloaded {
-				newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloaded
-				continue
-			}
-		}
-
-		jobName := modelInfo.ModelName + "-" + localModelNode.ObjectMeta.Name
-
-		job, err := c.launchJob(ctx, jobName, localModelNode, modelInfo, modelInfo.ModelName)
+		var job *batchv1.Job
+		folderExists, err := fsHelper.hasModelFolder(modelInfo.ModelName)
 		if err != nil {
-			c.Log.Error(err, "Job error", "name", jobName)
+			c.Log.Error(err, "Failed to check model folder", "model", modelInfo.ModelName)
 			return err
 		}
-
-		switch {
-		case job.Status.Succeeded > 0:
-			newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloaded
-		case job.Status.Failed > 0:
-			newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloadError
-		case job.Status.Ready != nil && *job.Status.Ready > 0:
-			newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloading
-		default:
-			newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloadPending
+		if folderExists {
+			c.Log.Info("Model folder found", "model", modelInfo.ModelName)
+			// If folder exists and the job has been successfully completed, do nothing
+			if status, ok := localModelNode.Status.ModelStatus[modelInfo.ModelName]; ok {
+				if status == v1alpha1api.ModelDownloaded {
+					newStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloaded
+				}
+				continue
+			}
+			job, err = c.getLatestJob(ctx, modelInfo.ModelName, nodeName, false)
+			if err != nil {
+				c.Log.Error(err, "Failed to getLatestJob", "model", modelInfo.ModelName, "node", nodeName)
+				return err
+			}
+			// If job is not found, create a new one. Because download could be incomplete.
+			if job == nil {
+				c.Log.Info("Model folder exists, creating download job", "model", modelInfo.ModelName)
+				job, err = c.launchJob(ctx, *localModelNode, modelInfo)
+				if err != nil {
+					c.Log.Error(err, "Failed to create Job", "model", modelInfo.ModelName, "node", nodeName)
+					return err
+				}
+			}
+		} else {
+			// Folder does not exist
+			c.Log.Info("Model folder not found", "model", modelInfo.ModelName)
+			job, err = c.getLatestJob(ctx, modelInfo.ModelName, nodeName, true)
+			if err != nil {
+				c.Log.Error(err, "Failed to getLatestJob", "model", modelInfo.ModelName, "node", nodeName)
+				return err
+			}
+			if job != nil {
+				c.Log.Info("model status from latest job", "model", modelInfo.ModelName, "status", getModelStatusFromJobStatus(job.Status))
+			}
+			// Recreate job if it has been terminated because the model is missing locally
+			if job == nil || job.Status.Succeeded > 0 {
```

Comment:
There could also be the case the job is failed without creating the folder.

Reply:
Yeah, then we do not retry these jobs until they are manually removed. What do you think?